R/data.table.R

```diff
@@ -1892,72 +1890,293 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
+      X.info$classes = X.info$classes[-cols.to.drop]
+      X.info$types = X.info$types[-cols.to.drop]
+    } else if (!missing(cols.to.update)) {
+      X.info$classes[cols.to.update] = lapply(X[cols.to.update], class)
+      X.info$types[cols.to.update] = vapply_1c(X[cols.to.update], typeof)
+    } else {
+      stop("Internal error: cols.to.update and cols.to.drop cannot both be supplied to column_properties()") # nocov
+    }
+    
+    # Get unique classes and types
+    X.info$uniq.class.list = unique(X.info$classes) # list, i.e. handles multi-class columns
+    X.info$uniq.classes = sort(unique(unlist(X.info$uniq.class.list))) # flat vector of unique classes
+    X.info$uniq.types = sort(unique(X.info$types))
+
+    X.info
+  }
+  X.info = column_properties(X)
+  
+  # Check whether any column has type or class among the 'target' vector,
+  # by matching to the precomputed vector of unique class or type information
+  # stored in X.info. col.uniqs should either be X.info$uniq.classes or 
+  # X.info$uniq.types. Returns a single TRUE or FALSE.
+  any_col_is = function(col.uniqs, target) { 
+    any(target %chin% col.uniqs)
+  }
+  
+  # Check whether all columns have type or class among the 'target' vector,
+  # by matching to the precomputed vector of unique class or type information
+  # stored in X.info. col.uniqs should either be X.info$uniq.classes or 
+  # X.info$uniq.types. Returns a single TRUE or FALSE.
+  all_col_is = function(col.uniqs, target) {
+    length(setdiff(col.uniqs, target)) == 0L
+  }
+  
+  # Check whether each column has type or class among the 'target' vector,
+  # by matching to the precomputed vector of class or type information
+  # stored in X.info. col.info should either be X.info$types or X.info$classes.
+  # Returns TRUE or FALSE for each column, allowing for which() or which(!()) to
+  # be subsequently applied to get column indices for subsetting. This function
+  # should generally be applied only after an any_col_is() or all_col_is() check
+  # to avoid expensive checks across all columns where they are not needed (i.e.
+  # if all columns are of a single type already we want to minimise the column
+  # check computation)
+  col_is = function(col.info, target) { 
+    if (length(target) > 1)
+      vapply_1b(col.info, function(xj) { any(target %chin% xj) } )
+    else
+      vapply_1b(col.info, `==`, target)
+  }
+  
+  # Check and fix the data.table if it is malformed - i.e. contains NULL
+  # columns, wide columns, or columns whose length != .N. For NULL columns
+  # we need to drop manually, otherwise we can use as.data.table().
+  if (any_col_is(X.info$uniq.types, "NULL")) {
+    col.is.null = which(col_is(X.info$types, "NULL"))
+    X[col.is.null] = NULL
+    X.info = column_properties(X, X.info, cols.to.drop = col.is.null) # update column properties info
+  }
+  
+  has.wide.columns = any(vapply_1b(X, function(xj) length(dim(xj)) > 0L)) 
+  n.column.lengths = length(unique(vapply_1i(X, length)))
```

Comment:
When the input data.table has many columns these two calls are quite expensive in the context of the pre column checks, as these attributes are not used for any other column checks. 

It would be nice if we could somehow avoid these, but I can't see how, as they're absolutely necessary for avoiding problems with memcpy in the C code (as the memcpy call assumes the number of elements to copy is .N - if this is not true bad things can happen). 

As far as I can tell, the only way these checks will fail is if a data.table has been created through `structure(list(...), class=c("data.table", "data.frame"))` as `data.table()`, `as.data.table()`, and `setDT()` will error if trying to create malformed columns like this. 

Is there a way we can check for a valid externalptr in `attributes(dt)[[".internal.selfref"]]` to avoid these expensive checks on valid data.tables?

Reply:
Have found the internal `selfrefok()` function. Unfortunately `selfrefok()` does not guarantee the data.table is not malformed, as `setDT()` will allow these columns and then `selfrefok()` will be `1L`. E.g. as in test `2074.07`. This is particularly problematic if one of the columns has length < .N.